services: add keycloak

gabyx opened the pull request:
- Add `keycloak` service.
- Add optional realm import and export.
- Add example.
- [x] Test that it works.
- [x] Make health prope script work.

@andrzejressel, @tboerger : I try to continue your effort to make it work.

Notes: 
I have a docker-compose file which works so far:
https://gitlab.com/data-custodian/custodian/-/blob/beea86076aef53ea00d7bf2aca7832d237022ac4/deploy/docker/docker-compose.yaml#L352

tboerger commented:
Looks already quite good, I will try to figure out a working configuration. Some more options like ports would be great and maybe even some env config to add more env variables.

gabyx commented:
@tboerger: Port is already there. Jeah Env config seems good!

gabyx commented:
@domenkozar: Maybe you could help me in understanding the test
I run `direnv allow`, I think its needed to build the rust stuff and include all modules?
then I ran `./result/bin/devenv-run-tests --only keycloak`
Keycloak seems to start, but I get a ` error: The option `services.keycloak' does not exist. Definition values`
error which is strange? Why is that? What do I miss. 

Thanks a lot

```shell
./result/bin/devenv-run-tests --only keycloak
Running Tests
Running in directory examples
Initialized empty Git repository in /tmp/devenv-run-tests-keycloak.uhfoSoeeAasj/.git/
  Running keycloak
• Building tests ...
• Using Cachix: devenv
these 2 derivations will be built:
  /nix/store/zhhddb78fk2nm6r9w6mrc8jwjlj42yx4-devenv-tasks-1.4.2.drv
  /nix/store/02dw5q7y2dwa0i9gl0ijp06sg6mgm7nw-devenv-test.drv
building '/nix/store/zhhddb78fk2nm6r9w6mrc8jwjlj42yx4-devenv-tasks-1.4.2.drv'...
building '/nix/store/02dw5q7y2dwa0i9gl0ijp06sg6mgm7nw-devenv-test.drv'...
✔ Building tests in 43.5s
• Building processes ...
these 4 derivations will be built:
  /nix/store/idm92b7dkxqz3r2vlybj0ccnlk6l5aap-start-keycloak.drv
  /nix/store/8xll3akhf160xwzl12jamj8w8y69lmhp-keycloak.drv
  /nix/store/bbs3mi9d1h9ziyn7p600293qaw0bf2ls-process-compose.yaml.drv
  /nix/store/jyk58cp557dmw4xf6yd450jl6ca1f76m-devenv-up.drv
building '/nix/store/idm92b7dkxqz3r2vlybj0ccnlk6l5aap-start-keycloak.drv'...
building '/nix/store/8xll3akhf160xwzl12jamj8w8y69lmhp-keycloak.drv'...
building '/nix/store/bbs3mi9d1h9ziyn7p600293qaw0bf2ls-process-compose.yaml.drv'...
building '/nix/store/jyk58cp557dmw4xf6yd450jl6ca1f76m-devenv-up.drv'...
✔ Building processes in 4.11s
• Starting processes ...
• Building shell ...
building '/nix/store/v7ssf93cyw0yfkxdjrhfp6srlygayqfz-devenv-shell-env.drv'...
✔ Building shell in 5.39s
warning: not writing modified lock file of flake 'git+file:///tmp/devenv-run-tests-keycloak.uhfoSoeeAasj':
• Added input 'devenv':
    'path:/persist/repos/devenv?dir=src/modules'
• Added input 'git-hooks':
    'github:cachix/git-hooks.nix/dcf5072734cb576d2b0c59b2ac44f5050b5eac82' (2025-03-22)
• Added input 'git-hooks/flake-compat':
    'github:edolstra/flake-compat/ff81ac966bb2cae68946d5ed5fc4994f96d0ffec' (2024-12-04)
• Added input 'git-hooks/gitignore':
    'github:hercules-ci/gitignore.nix/637db329424fd7e46cf4185293b9cc8c88c95394' (2024-02-28)
• Added input 'git-hooks/gitignore/nixpkgs':
    follows 'git-hooks/nixpkgs'
• Added input 'git-hooks/nixpkgs':
    follows 'nixpkgs'
• Added input 'nixpkgs':
    'github:cachix/devenv-nixpkgs/7bd9e84d0452f6d2e63b6e6da29fe73fac951857' (2024-12-06)
• Added input 'pre-commit-hooks':
    follows 'git-hooks'
PID is 1125694
Stop:      $ devenv processes stop
✔ Starting processes in 5.98s
• Running tests ...
• Building shell ...
Running tasks     devenv:enterShell
Succeeded         devenv:enterShell 3ms
1 Succeeded                         3.44ms

✨ devenv 1.4.2 is newer than devenv input (1.4.1) in devenv.lock. Run 'devenv update' to sync.
[keycloak       ] Current Mode: production
[keycloak       ] Current Configuration:
[keycloak       ]       kc.bootstrap-admin-password =  ******* (ENV)
[keycloak       ]       kc.bootstrap-admin-username =  admin (ENV)
[keycloak       ]       kc.conf-dir =  /tmp/devenv-run-tests-keycloak.uhfoSoeeAasj/.devenv/state/keycloak/conf (ENV)
[keycloak       ]       kc.config.args =  show-config (SysPropConfigSource)
[keycloak       ]       kc.config.built =  true (SysPropConfigSource)
[keycloak       ]       kc.db =  dev-mem (ENV)
[keycloak       ]       kc.health-enable =  true (ENV)
[keycloak       ]       kc.home-dir =  /tmp/devenv-run-tests-keycloak.uhfoSoeeAasj/.devenv/state/keycloak (ENV)
[keycloak       ]       kc.hostname =  localhost (ENV)
[keycloak       ]       kc.http-port =  8080 (ENV)
[keycloak       ]       kc.log =  console (ENV)
[keycloak       ]       kc.log-console-color =  true (ENV)
[keycloak       ]       kc.log-console-output =  default (classpath keycloak.conf)
[keycloak       ]       kc.log-level =  debug (ENV)
[keycloak       ]       kc.spi-hostname-v2-hostname =  localhost (ENV)
[keycloak       ]       kc.tmp-dir =  /tmp/devenv-run-tests-keycloak.uhfoSoeeAasj/.devenv/state/keycloak/tmp (ENV)
[keycloak       ]       kc.version =  26.0.6 (SysPropConfigSource)
[keycloak       ] Updating the configuration and installing your custom providers, if any. Please wait.
✔ Building shell in 4.35s
warning: not writing modified lock file of flake 'git+file:///tmp/devenv-run-tests-keycloak.uhfoSoeeAasj':
• Added input 'devenv':
    'path:/persist/repos/devenv?dir=src/modules'
• Added input 'git-hooks':
    'github:cachix/git-hooks.nix/dcf5072734cb576d2b0c59b2ac44f5050b5eac82' (2025-03-22)
• Added input 'git-hooks/flake-compat':
    'github:edolstra/flake-compat/ff81ac966bb2cae68946d5ed5fc4994f96d0ffec' (2024-12-04)
• Added input 'git-hooks/gitignore':
    'github:hercules-ci/gitignore.nix/637db329424fd7e46cf4185293b9cc8c88c95394' (2024-02-28)
• Added input 'git-hooks/gitignore/nixpkgs':
    follows 'git-hooks/nixpkgs'
• Added input 'git-hooks/nixpkgs':
    follows 'nixpkgs'
• Added input 'nixpkgs':
    'github:cachix/devenv-nixpkgs/7bd9e84d0452f6d2e63b6e6da29fe73fac951857' (2024-12-06)
• Added input 'pre-commit-hooks':
    follows 'git-hooks'
• Executing in shell ...
Running tasks     devenv:enterShell
Succeeded         devenv:enterShell 2ms
1 Succeeded                         3.06ms

✨ devenv 1.4.2 is newer than devenv input (1.4.1) in devenv.lock. Run 'devenv update' to sync.
• Setting up shell environment ...
Running tasks     devenv:enterShell
Succeeded         devenv:enterShell 3ms
1 Succeeded                         3.33ms

✨ devenv 1.4.2 is newer than devenv input (1.4.1) in devenv.lock. Run 'devenv update' to sync.
• Testing ...
Running tasks     devenv:enterTest
Not implemented   devenv:enterTest
1 Skipped                           23.69µs

• Running .test.sh...
+ devenv up
warning: creating lock file '/tmp/devenv-run-tests-keycloak.uhfoSoeeAasj/devenv.lock'
error:
       … while evaluating the attribute 'config'
         at «github:cachix/devenv-nixpkgs/7bd9e84d0452f6d2e63b6e6da29fe73fac951857»/lib/modules.nix:334:9:
          333|         options = checked options;
          334|         config = checked (removeAttrs config [ "_module" ]);
             |         ^
          335|         _module = checked (config._module);

       … while calling the 'seq' builtin
         at «github:cachix/devenv-nixpkgs/7bd9e84d0452f6d2e63b6e6da29fe73fac951857»/lib/modules.nix:334:18:
          333|         options = checked options;
          334|         config = checked (removeAttrs config [ "_module" ]);
             |                  ^
          335|         _module = checked (config._module);

       … while calling the 'throw' builtin
         at «github:cachix/devenv-nixpkgs/7bd9e84d0452f6d2e63b6e6da29fe73fac951857»/lib/modules.nix:310:18:
          309|                     ''
          310|             else throw baseMsg
             |                  ^
          311|         else null;

       error:
       error: The option `services.keycloak' does not exist. Definition values:
       - In `/tmp/devenv-run-tests-keycloak.uhfoSoeeAasj/devenv.nix':
           {
             enable = true;
           }
Error:   x Command `/nix/store/ni24hs509kjhi1ckm6wxhhj8h333437l-nix-devenv-2.24.0pre20250312_f3f44b2/bin/nix --show-trace --extra-
  | experimental-features nix-command --extra-experimental-features flakes --option warn-dirty false --keep-going --max-jobs
  | 16 --option eval-cache false eval --json .#devenv.processes -vv --log-format internal-json` failed with with exit code 1

✔ Executing in shell in 613ms
✔ Running tests in 5.64s
Stopping process with PID 1125694
Tests failed :(
Running in directory tests
Stopping processes...

keycloak: Failed

Ran 1 tests, 1 failed.
Error: Some tests failed
```

gabyx commented:
I think I cannot make it work. Keycloak tries to write (I assume) (it builds itself?) to the Nix store when launched, or to another write only directory.

I am not so familiar what the `"$exeDir/bin/kc.sh" --verbose start-dev` actually does.

I always get: some dumb JAVA build failures.

```shell
[keycloak       ] ERROR: Failed to run 'build' command.
[keycloak       ] ERROR: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[keycloak       ]       [error]: Build step io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar threw an exception: java.nio.file.ReadOnlyFileSystemException
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.checkWritable(ZipFileSystem.java:370)
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.createDirectory(ZipFileSystem.java:708)
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipPath.createDirectory(ZipPath.java:742)
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.createDirectory(ZipFileSystemProvider.java:186)
[keycloak       ]       at java.base/java.nio.file.Files.createDirectory(Files.java:699)
[keycloak       ]       at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:807)
[keycloak       ]       at java.base/java.nio.file.Files.createDirectories(Files.java:752)
[keycloak       ]       at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildThinJar(JarResultBuildStep.java:664)
[keycloak       ]       at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildRunnerJar(JarResultBuildStep.java:228)
[keycloak       ]       at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[keycloak       ]       at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[keycloak       ]       at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[keycloak       ]       at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[keycloak       ]       at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[keycloak       ]       at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[keycloak       ]       at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[keycloak       ]       at java.base/java.lang.Thread.run(Thread.java:1583)
[keycloak       ]       at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[keycloak       ]
[keycloak       ] ERROR: Build failure: Build failed due to errors
[keycloak       ]       [error]: Build step io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar threw an exception: java.nio.file.ReadOnlyFileSystemException
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.checkWritable(ZipFileSystem.java:370)
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.createDirectory(ZipFileSystem.java:708)
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipPath.createDirectory(ZipPath.java:742)
[keycloak       ]       at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.createDirectory(ZipFileSystemProvider.java:186)
[keycloak       ]       at java.base/java.nio.file.Files.createDirectory(Files.java:699)
[keycloak       ]       at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:807)
[keycloak       ]       at java.base/java.nio.file.Files.createDirectories(Files.java:752)
[keycloak       ]       at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildThinJar(JarResultBuildStep.java:664)
[keycloak       ]       at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildRunnerJar(JarResultBuildStep.java:228)
[keycloak       ]       at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[keycloak       ]       at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[keycloak       ]       at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[keycloak       ]       at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[keycloak       ]       at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[keycloak       ]       at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[keycloak       ]       at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[keycloak       ]       at java.base/java.lang.Thread.run(Thread.java:1583)
[keycloak       ]       at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[keycloak       ]
[keycloak       ] For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
```

@talyz: Do you might know what the problem is here since you have hopefully some insight. Does the NixOS modules do  something different? https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/keycloak.nix#L766

I just try to launch it in `http` only mode: https://github.com/cachix/devenv/pull/1806/files#diff-d8696b88eb47713cf77037d6a38f2ddeb99a11d2256064160d2aaa67ca6b36e1R133

tboerger commented:
Have you tried start instead if start-dev?

talyz commented:
Yes, you do indeed need to provide the config as is done in the NixOS module, since quarkus will try to rebuild its image if the config doesn't match its cached one.

gabyx commented:
@sandydoo: Is is actually possible to reuse (not sure how?) NixOS Modules and is the module system of devenv the same as NixOS's one?
Or should we just copy&pasta&adapt the stuff for a devenv service, probably because we do not want certain things like in the real service etc? Whats the advice here?

@tboerger: I will first try to make `dev-mem` db settings work, and then when we have a working thing, we could add the postgress stuff maybe by enabling automatically the prostgres service and using this instance etc... but slowely.

gabyx commented:
@talyz: Kudos for your module, its great!

tboerger commented:
Since this is for dev environments I would prefer to keep it simple without postgres and just use the builtin file-based database. Was it sqlite or h2 by default?

gabyx commented:
not sure I think its dev-mem or dev-file you can choose from

tboerger commented:
If you simply skip the env variable for the database driver it should be fine.

gabyx commented:
@tboerger: I dont have time next week to test if it starts up. So far the thing should build, maybe you can give it a test,
```bash
direnv allow # or nix develop 
./result/bin/devenv-run-tests --only keycloak
```
 if it starts up or not. Maybe propose changes I can integrate after 14. April.
 I added also (not tested yet) a process which exports realms, really needed to make changes and keep it presistent.

gabyx commented:
@talyz: Do you mind maybe helping out with some questions: I was hoping to get `keycloak` up and running with `db = dev-file` or `dev-mem` but I guess these option are always overridden by the derivation build. Do you know how I can build it with `dev-mem` or `dev-file`? 
I cannot start it unfortunately with `kc.sh start-dev` cause that always builds the stuff (?).
I see that keycloak passes a `--profile=dev` flag in https://github.com/keycloak/keycloak/blob/main/quarkus/dist/src/main/content/bin/kc.sh#L70
Probably I need to pass this to the build ?

gabyx commented:
> @talyz: Do you mind maybe helping out with some questions: I was hoping to get `keycloak` up and running with `db = dev-file` or `dev-mem` but I guess these option are always overridden by the derivation build. Do you know how I can build it with `dev-mem` or `dev-file`? 
> 
> I cannot start it unfortunately with `kc.sh start-dev` cause that always builds the stuff (?).
> 
> I see that keycloak passes a `--profile=dev` flag in https://github.com/keycloak/keycloak/blob/main/quarkus/dist/src/main/content/bin/kc.sh#L70
> 
> Probably I need to pass this to the build ?
> 
> 

It works now: I forgot to symlink some folders which are apparently used for the build, that was quite a hassle. I try to make the export/import of the realms also work...

tboerger commented on src/modules/services/keycloak.nix:
```diff
@@ -0,0 +1,493 @@
+
+    database = {
+      type = mkOption {
+        type = types.enum [
```
Since it would also work with postgres or mariadb does this enum really make sense? By default it's anyway using a file-based database which should be stored in the devenv state path.

gabyx replied:
Yes that includes secrets and everything. Thats exactly what we want as well, so everything is deterministic if you reimport it.

gabyx replied:
so far we only support `dev-mem` and `dev-file` thats why I only added those, the otherones need more complex setup etc etc. Does that make sense?

tboerger reviewed (commented):
I haven't checked the args of the export command, but does this also include secrets? That would help to have some reproducible setup for the development of apps against keycloak.

gabyx commented:
> I haven't checked the args of the export command, but does this also include secrets? That would help to have some reproducible setup for the development of apps against keycloak.

Yes that includes secrets and everything. Thats exactly what we want as well, so everything is deterministic if you reimport it.

gabyx commented:
@domenkozar: How can I request a review from maintainers? Would be really good, because the feature would help already to be in devenv. =)

domenkozar commented on src/modules/services/keycloak.nix:
```diff
@@ -0,0 +1,498 @@
+in
+{
+  imports = [
+    (mkRenamedOptionModule [ "keycloak" "enable" ] [ "services" "keycloak" "enable" ])
```
Not sure I understand why this is needed, `keycloak.enable` did not exist before this PR?

gabyx replied:
:+1: : Thanks thats a mistake from NixOS pkgs!

domenkozar commented on src/modules/services/keycloak.nix:
```diff
@@ -0,0 +1,498 @@
+    };
+
+    sslCertificate = mkOption {
+      type = nullOr types.path;
```
```suggestion
      type = nullOr (types.pathWith { inStore = false; absolute = true; });
```

This gets rid of `assertStringPath` helper

gabyx commented:
@domenkozar: Do you know where `types.pathWith` comes from. I rebased to main, but the `lib.types.pathWith` does somehow not exist yet in `devenv-nixpkgs`. Is it new I guess.

I guess `devenv-nixpkgs` has this function not yet integrated, correct?

```
 … while evaluating the attribute 'types.pathWith'
         at «github:cachix/devenv-nixpkgs/7bd9e84d0452f6d2e63b6e6da29fe73fac951857»/lib/default.nix:35:5:
           34|     options = callLibs ./options.nix;
           35|     types = callLibs ./types.nix;
             |     ^
           36|

       error: at /persist/repos/devenv/src/modules/services/keycloak.nix:40:9:
           39|       type = nullOr (
           40|         lib.types.pathWith {
             |         ^
           41|           inStore = false;

       error: attribute 'pathWith' missing
 ```

gabyx commented:
@domenkozar: Do I need to wait till the updates are on `main`?

tboerger commented:
I would like to integrate this module into my flake.nix based devenv environment, does somebody how I can do this? It looks like this PR will take some more time to get merged.

gabyx commented:
I think its ready! its just that some functions are missing. I try to rebase to main on Monday again.

gabyx commented:
@tboerger, @domenkozar: The PR has been rebase, and another test was added for `sslCertificates`.
Thanks for merging. (Check the last two commits).

domenkozar commented:
Thank you for all this work @gabyx :) Due to GitHub Actions limitations we've hit 250 jobs limit, could I ask you to move keycloack from examples to tests? We're going to fix this in the future but for now this is the only option we have.

gabyx commented:
Jeah, I try. If I understand them =).

sandydoo commented on src/modules/services/keycloak.nix:
```diff
@@ -0,0 +1,491 @@
+        KC_TMP_DIR = config.env.DEVENV_STATE + "/keycloak/tmp";
+
+        KC_BOOTSTRAP_ADMIN_PASSWORD = "admin";
+        KC_BOOTSTRAP_ADMIN_USERNAME = "${lib.escapeShellArg cfg.initialAdminPassword}";
```
https://github.com/cachix/devenv/pull/1939

gabyx commented:
I can add a fix for that, check tomorrow probably a simple switch.

tboerger commented:
The twisted username/password fix is already merged. But an option to hide the export process within the process manager would be great :)
